rlp/raw.go

```diff
@@ -292,3 +286,52 @@ func AppendUint64(b []byte, i uint64) []byte {
 		)
 	}
 }
+
+// AppendUint64 appends the RLP encoding of i to b, and returns the resulting slice.
+func AppendUint64(b []byte, i uint64) []byte {
+	if i == 0 {
+		return append(b, 0x80)
+	} else if i < 128 {
+		// fits single byte
+		return append(b, byte(i))
+	} else {
+		return appendUintWithTag(b, i, 0x80)
+	}
+}
+
+// AppendString appends RLP-encoded str to buf and returns the extended buffer.
+func AppendString(buf, str []byte) []byte {
+	if size := len(str); size == 0 {
+		return append(buf, 0x80)
+	} else if size == 1 && str[0] < 128 {
+		// fits single byte, no string header
+		return append(buf, str[0])
+	} else if size < 56 {
+		buf = append(buf, 0x80+byte(size))
+		return append(buf, str...)
+	} else {
+		buf = appendUintWithTag(buf, uint64(size), 0xB7)
+		return append(buf, str...)
+	}
+}
+
+// EndList ends up list starting from offset and returns the extended buffer.
+// Content after offset is treated as list content.
+func EndList(buf []byte, offset int) []byte {
+	contentSize := len(buf) - offset
+	if contentSize == 0 {
+		buf = append(buf, 0xC0)
+	} else if contentSize < 56 {
+		// shift the content for room of list header
+		buf = append(buf[:offset+1], buf[offset:]...)
+		// write list header
+		buf[offset] = 0xC0 + byte(contentSize)
+	} else {
+		headerSize := intsize(uint64(contentSize)) + 1
+		// shift the content for room of list header
+		buf = append(buf[:offset+headerSize], buf[offset:]...)
+		// write list header
+		appendUintWithTag(buf[:offset], uint64(contentSize), 0xF7)
```

Comment:
Ah, wait, my bad. You are actually _intentionally_ ignoring the returnvalue here, because you're not _actually_ appending, you are creating a new buffer which aliases the old one, and you are using the append-functionality to just overwrite into the old header. 

It's a neat trick, but I don't like it. It *looks* like a bug, and might be "fixed" by someone later on.

Reply:
We can leave it as is, but please add a comment explaining why it does what it does